### ISSUE TITLE
JSON percent-cpu-time gos down with cpu sampling rate and lack of time absolute metrics

### ISSUE THREAD

**Issue by fjgmoya** [plasma-umass/scalene#569]
**Thank you so much** for this scalene smart profiler. 
Profiler produces a --json file describing CPU and eventually GPU time consumption.

Hower, I can see **overall percent cpu time decreases when cpu-sampling-rate goes down**. This is my code sample and collected results from this case:

```
import numpy as np
import smote_variants as sv
from sklearn.datasets import make_classification
import scalene

def profile_me():
    X, y = make_classification(n_samples=100000, n_features=99, random_state=13)
    smote = sv.SMOTE(n_neighbors=5, random_state=13, n_jobs=1)
    X_smoted, Y_smoted = smote.sample(X, y)

%scrun --outfile profile.json --cli --json --cpu-only --cpu-sampling-rate=0.01 profile_me()
```
![datos_profile_cpu](https://user-images.githubusercontent.com/38583999/218254060-202c79a7-c0ef-4eb4-856e-d8a1a8bc95e3.JPG)

When using a GPU-oriented sample:

```
import numpy as np
from sklearn.datasets import make_classification
import scalene
import faiss

def profile_gpu_me(n_neighbors=5, n_gpu=0):
    X, y = make_classification(n_samples=100000, n_features=99, random_state=13)
    res=faiss.StandardGpuResources()
    index = faiss.IndexFlatL2(X.shape[1])
    gpu_index_flat = faiss.index_cpu_to_gpu(res, n_gpu, index)
    gpu_index_flat.add(X.astype(np.float32))
    distances, indices = gpu_index_flat.search(X.astype(np.float32), int(n_neighbors))

%scrun --outfile profile.json --cli --json --cpu-only --cpu-sampling-rate=0.01 profile_gpu_me()
```
And summarized results:
![datos_profile_gpu](https://user-images.githubusercontent.com/38583999/218254135-c012d849-83e5-4760-a86c-7192bfad385c.JPG)

With these simple cases and another others, CPU percentage decreases, I can't obtaon a 100% summarization, whereas GPU time always offers a 100% sum.

![image](https://user-images.githubusercontent.com/38583999/218254394-a78451d0-b88f-451a-8c8f-b03ea9bace84.png)

**Two main questions about this:**

1. How can I obtain a 100% CPU summarization time? Is something else I could do?  I notice some time is loss, but I can't assignt this to system python or others. In other hard-enough experiments, cpu loss percentage climbs over 50%. Profile only item is not used in order to  geta  general focus. 
2. JSON files offers an alone absolute time measure (elapsed_time_sec), but the rest of time measures are percent ones. As shown, if elapsed time is 2.725770473480224, my cpu-percent time is 91.7504550175195 and n_gpu_percent is 99.99999999, how can I compute real CPU and GPU time as absolute time measures in seconds? Is 2.725770473480224 only CPU time or sums CPU+GPU time?

Thanks in advance.

About smote-variants:
https://github.com/analyticalmindsltd/smote_variants

About faiss:
https://github.com/facebookresearch/faiss

**Comment by emeryberger** [comment]
I'm not clear why you are increasing the sampling rate, but it makes perfect sense that this distorts the results. 

Each sample takes a fixed amount of time to process. If you take samples too frequently, then Scalene will end up spending a significant fraction of time taking samples. At some point, the sampling rate will get so high that the act of sampling will produce too much overhead. The current, default sampling rate should provide high accuracy with little distortion (we have empirically verified this).

**Comment by fjgmoya** [comment]
Thank you for your so light-speed response, Emery:

Yes, I thought so. The only reason for me to increasy sampling rate parameter is get a nearly 100% of computed CPU time, as obtained with GPU. No matter about 99% or similar percentages, but I made another tests wich offered something like a 50%, and I tried to get a better computed time response using (wrongly) sampling rate .

The second question I asked is a main one: who to compute absolute execution time from CPU and GPU where an only elapse-time measure is available from JSON file. 

Thank you so much for your support, and congrats for this scalene profiler.

**Comment by emeryberger** [comment]
Can you explain why you need absolute execution time numbers? I don't see why you need more than the utilization numbers. FWIW the latest version in the repo now adds % core utilization. Sequential Python code can of course never use more than one core effectively, but native libraries should be able to do so (e.g., `numpy`).

**Comment by fjgmoya** [comment]
Thanks, Emery:

We're tasting some algorithms with both CPU and GPU versions. In case of GPU one, there is time consumption on CPU and GPU too, so we need to get separtely absolute GPU and CPU times. There is another main reason: a well-argued comparative from both CPU and GPU version of the same algorith requiere absolute CPU and GPU measures, as you see, in order to obtain the overall gaining from the usage from GPU one to CPU version.

Thanks in advance.
